crates/interpreter/src/exec.rs

```diff
@@ -1061,14 +1061,15 @@ impl<'m> Thread<'m> {
         let frame = self.frame();
         let label = frame.labels.pop().unwrap();
         if frame.labels.is_empty() {
-            let frame = self.frames.pop().unwrap();
-            debug_assert_eq!(frame.labels_values.len(), label.values_cnt);
-            debug_assert_eq!(frame.labels_values.len(), frame.arity);
+            let mut frame = self.frames.pop().unwrap();
+            frame.values.drain(0 .. frame.locals_cnt);
+            debug_assert_eq!(frame.values.len(), label.values_cnt);
+            debug_assert_eq!(frame.values.len(), frame.arity);
             if self.frames.is_empty() {
-                return ThreadResult::Done(frame.labels_values);
+                return ThreadResult::Done(frame.values);
             }
             unsafe { self.parser.restore(frame.ret) };
-            self.values().extend(frame.labels_values);
+            self.values().extend(frame.values);
```

Comment:
@ia0 You mentioned earlier that this line could be optimized. Here `self.values()` refers to the values of the frame before the popped `frame`, and it does not know about `frame.values`. Could you explain more? Thanks.

Reply:
This will get optimized once we use a single value stack for the whole thread (not just per function frame). But this will only be true if we go with the "keep locals outside the stack" option (discussed above). Otherwise we'll have to drain.